Fix issues found by  golangci-lint

alanconway opened the pull request:
golangcli-lint combines multiple linters, including go vet. It has
good default settings - not much noise, I recommend it. Fixes are to:

- Remove dead/unreachable code
- Fix missing error checks:
  I mostly marked them intentional with `_ = foo()`, I think this is right
  One was in in a `defer` statement I fixed by returning it.
- Minor simplifications (from 'gosimple' linter)

n3wscott commented on pkg/cloudevents/observability/observer.go:
```diff
@@ -27,12 +27,11 @@ type Reporter interface {
-	span    *trace.Span
-	on      Observable
-	start   time.Time
-	measure stats.Measure
```
`measure stats.Measure` - I am suppressed this is not used. I think we need the measurement.

n3wscott replied:
ah, looks like it was a migration leftovers. 👍

n3wscott commented on pkg/cloudevents/transport/amqp/codec_test.go:
```diff
@@ -370,13 +370,6 @@ func TestCodecRoundTrip(t *testing.T) {
 					got.Data = data
 				}
 
-				if tc.wantErr != nil || err != nil {
```
the tests need to be expanded to non-happy path only. Please leave this?

n3wscott commented on pkg/cloudevents/transport/amqp/codec_test.go:
```diff
@@ -508,13 +501,6 @@ func TestCodecAsMiddleware(t *testing.T) {
 					got.Data = data
 				}
 
-				if tc.wantErr != nil || err != nil {
```
the tests need to be expanded to non-happy path only. Please leave this?

alanconway replied:
will do, although right now it is unreachable.

n3wscott commented on pkg/cloudevents/transport/http/codec_test.go:
```diff
@@ -797,13 +797,6 @@ func TestCodecRoundTrip(t *testing.T) {
 					got.Data = data
 				}
 
-				if tc.wantErr != nil || err != nil {
```
the tests need to be expanded to non-happy path only. Please leave this?

alanconway replied:
will do, although right now it is unreachable

n3wscott commented on pkg/cloudevents/transport/http/codec_test.go:
```diff
@@ -939,13 +932,6 @@ func TestCodecAsMiddleware(t *testing.T) {
 						got.Data = data
 					}
 
-					if tc.wantErr != nil || err != nil {
```
the tests need to be expanded to non-happy path only. Please leave this?

n3wscott commented on pkg/cloudevents/transport/http/transport_test.go:
```diff
@@ -136,7 +136,7 @@ func TestStableConnectionsToSingleHost(t *testing.T) {
 func TestMiddleware(t *testing.T) {
 	testCases := map[string]struct {
 		middleware []string
-		want       string
```
hm that is a weird change.

alanconway replied:
Very! The want member is not used, I've no idea why I replaced it with "_" instead of just removing it. Removed.

n3wscott commented on pkg/cloudevents/transport/nats/codec_test.go:
```diff
@@ -240,13 +236,6 @@ func TestCodecRoundTrip(t *testing.T) {
 					got.Data = data
 				}
 
-				if tc.wantErr != nil || err != nil {
```
the tests need to be expanded to non-happy path only. Please leave this?

alanconway replied:
ok

n3wscott commented on pkg/cloudevents/transport/nats/codec_test.go:
```diff
@@ -378,13 +367,6 @@ func TestCodecAsMiddleware(t *testing.T) {
 					got.Data = data
 				}
 
-				if tc.wantErr != nil || err != nil {
```
the tests need to be expanded to non-happy path only. Please leave this?

alanconway replied:
ok

n3wscott commented on pkg/cloudevents/transport/nats/transport.go:
```diff
@@ -112,7 +112,7 @@ func (t *Transport) HasConverter() bool {
 // StartReceiver implements Transport.StartReceiver
 // NOTE: This is a blocking call.
-func (t *Transport) StartReceiver(ctx context.Context) error {
+func (t *Transport) StartReceiver(ctx context.Context) (err error) {
```
do not name the return params.

alanconway replied:
This is necessary to  fix your error return problem at the end of the function. You had a comment:
    // TODO: create an error channel to pass this up
A channel is not necessary. A deferred function executes within the scope of the function that declares it, and can set the return values - but only if they are named.

n3wscott replied:
Is that right?! Well thanks for that info. Great to know. LGTM

alanconway replied:
Also after a bit of googling I can't find any objections to named returns, and some articles on how they can be helpful: https://golang.org/doc/effective_go.html#named-results

IMO the only danger is the "naked" return, so I don't use it. I always list the return values even if they are the same variables that were named in parameters to avoid shadowing confusion, e.g.

```
func New() (t *Thing, err error) {
  if err := something(); err != nil {
     return // BAD 
     // naked return uses the variables which are still  (nil, nil)
     // The `err :=`  declaration shadows the err result.

     return t, err // GOOD
     // Explicit return evaluated in local context using inner err variable.
     // values are copied to the result variables before doing defers
 }
 ...
}
```

n3wscott commented on pkg/cloudevents/types/urlref.go:
```diff
@@ -51,7 +51,7 @@ func (u *URLRef) UnmarshalJSON(b []byte) error {
 func (u URLRef) MarshalXML(e *xml.Encoder, start xml.StartElement) error {
-	v := fmt.Sprintf("%s", u.String())
+	v := u.String()
 	return e.EncodeElement(v, start)
```
perhaps `return e.EncodeElement(u.String(), start)`

alanconway replied:
Yes, better.

n3wscott commented on test/http/loopback.go:
```diff
@@ -110,44 +110,3 @@ func ClientLoopback(t *testing.T, tc TapTest, topts ...cehttp.Option) {
 }
-
-// To help with debug, if needed.
-func printTap(t *testing.T, tap *tapHandler, testID string) {
```
please leave this, it was really helpful for debugging.

alanconway replied:
ok

n3wscott commented on test/http/tap_handler.go:
```diff
@@ -87,7 +87,7 @@ func (t *tapHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 
 func copyHeaders(from http.Header) http.Header {
 	to := http.Header{}
-	if from == nil || to == nil {
```
this will cause a npe

alanconway commented:
Fixed all requests except named return value required to return an error from a deferred function call.
Rebased and added some more fixes picked up from master - I think they are all uncontroversial

n3wscott commented:
Looks like you need to run `go fmt` 

Otherwise, LGTM, thanks for the work!

alanconway commented:
Fixed formatting, squashed into a single commit.

n3wscott commented:
LGTM! THANKS!!!
